Add ProcesorID, User TIme, Kernel Time to event metadata

tekert opened the pull request:
Adds missing fields to event header
ProcesorID, User TIme, Kernel Time

Fixes #4

tekert committed:
Add Missing Procesor, user and kernel times.

tekert committed:
Expose the members (alternative version)
And fix int64 to uint64

I think it's best since we are using public mode.
TODO: private mode.

tekert committed:
identation

qjerome commented:
Hi @tekert !

Thanks a lot for this PR. Have you tested that it is working as expected ?

tekert committed:
Final fix

tekert commented:
Im testing it since this morning, the thing is, i checked the code, it works, but i have no idea if the numbers are correct, i was trying to think how can i make UserTime move up to see if it works,
Im comparing the results with logman and tracerpt file.etl, give me some hours to see if it works, the thing is, at least code wise, bit wise, it should.

tekert commented:
Some preliminary results:

logman 2024-12-22T15:18:15.965880700"
KernelTime="705" 
UserTime="0" 
        
etw-go 2024-12-22T15:18:15.9658804Z
"KernelTime": 0,
"UserTime": 47,
"ProcessorTime": 47  (added for testing field that prints the uint64)
      
        

logman:
KernelTime =      0010 1100 0001
UserTime =   

etw-go:
ProcessorTime =        0010 1111     

No correlation.

The code seems ok, i don't know why those fields are different, for this event, everything else is equal bit by bit.
I'm going to dig deeper, maybe some struct is not aligned?

tekert commented:
Well, i could not find why the two traces cpu times are different, 
The values in `EventHeader` struct are correctly aligned, the other related structs too, The values before this field, and after, are also correctly populated, so.. I really don't know but those numbers are what the callback reports.

tekert commented:
Uhm, today found out that the numbers change only when they change in logman (had to sync the two runs in the dumps), so i think this is a problem of units (maybe), im getting closer i hope.

tekert commented:
Yes, confirmed, logman has every Time multipled exactly by 15 in my case.
Also the fields are swapped (uhmm)
edit: oh, forgot we used a uint64.. so the ordering is fine, the memory is laid in little endian order when the cpu writes 2 32bit chunks to it. Don't know the details of the go compiler but at least it works.

I will try to find why etw-go spits time metrics / 15

tekert committed:
Swap field memory order Kernel|User Time.

- Get UserTime from the most significat side and Kernel from the lower side.
Weird, struct fields are usually laid in order

- Fix TdhContext.ParameterValue uint32 to unit64 to match the C interface size.

tekert commented:
I did a test in C++, using the examples from https://learn.microsoft.com/en-us/windows/win32/etw/retrieving-event-metadata
added output for UserTime and KernelTime from `pEvent->EventHeader.KernelTime` and they exactly match the ones from this PR, so.. seems logman has a serious bug here. This PR can be safely merged.

qjerome commented:
Awesome, thank you for your work !
I'll also give it a try before merging but I'm not sure I'll be able to do it this week...

tekert commented:
No problem, me too im going out, in mid january i may have some other PR to get the length and count of properties from the fields instead of using old thd syscalls, it should help the performance grately according to the microsoft docs for non WPP events.
Happy new year!

qjerome commented:
Hey @tekert !

Sorry I completely forgot about this PR ... The fact you closed it reminded me it was open >< !
Anyway you've made a fork with lots of changes now and I guess you work with that :)

Thanks for your work :+1:

tekert commented:
Hey @qjerome , did not close it, but somehow changed the names of the some branches and that seems to have cut the link of the PR, yeah, i began adding some stuf for running this 24/7 on some machines with very low overhead, and one thing lead to another to another heh, but i almost finished.
